File: backend/src/pipeline.py
"""Orchestration : Open-Meteo AROME 0.025°, cache JSON, profil Stüve."""

from __future__ import annotations

import logging
import math
import time
from datetime import datetime, timezone
from typing import Any

from src.cache import PointCache
from src.config import (
    TIMEZONE,
    Settings,
    in_arome_domain,
    load_settings,
    round_point,
)
from src.openmeteo import HEIGHT_AGL_M, OM_GRID, OM_MODEL, PRESSURE_HPA, OpenMeteoClient
from src.profile import (
    CLOUD_BASE_RULE,
    HourBlock,
    ProfilePoint,
    SurfaceBlock,
    estimate_cloud_base,
    fuse_profile,
)

logger = logging.getLogger(__name__)


def _f(value: Any) -> float | None:
    if value is None:
        return None
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    if math.isnan(number) or math.isinf(number):
        return None
    return number


def _r1(x: float | None) -> float | None:
    return None if x is None else round(x, 1)


def _ri(x: float | None) -> int | None:
    return None if x is None else int(round(x))


def cache_slot_utc(now: datetime | None = None) -> str:
    """Slot 3 h (cycle AROME), pour invalider le cache — pas l'heure d'init MF."""
    utc = (now or datetime.now(timezone.utc)).astimezone(timezone.utc)
    hour = (utc.hour // 3) * 3
    return utc.replace(hour=hour, minute=0, second=0, microsecond=0).strftime(
        "%Y-%m-%dT%H:%M:%SZ"
    )


class AromePipeline:
    def __init__(self, settings: Settings | None = None) -> None:
        self.settings = settings or load_settings()
        self.om = OpenMeteoClient()
        self.cache = PointCache(self.settings.cache_dir)

    def forecast(
        self,
        lat: float,
        lon: float,
        *,
        time_from: datetime | None = None,
        time_to: datetime | None = None,
        force: bool = False,
    ) -> dict[str, Any]:
        if not in_arome_domain(lat, lon):
            raise ValueError(
                f"Point ({lat}, {lon}) hors domaine AROME "
                f"lat 37.5–55.4 N, lon 12 W–16 E"
            )
        rlat, rlon = round_point(lat, lon)
        window_from, window_to = resolve_window(time_from, time_to)
        slot = cache_slot_utc()
        wf = window_from.isoformat()
        wt = window_to.isoformat()

        t0 = time.perf_counter()
        if not force:
            cached = self.cache.read(rlat, rlon, slot, wf, wt)
            if cached is not None:
                logger.info("cache hit Open-Meteo %.0f ms", (time.perf_counter() - t0) * 1000)
                return cached

        lock = self.cache.lock(rlat, rlon, slot, wf, wt)
        with lock:
            if not force:
                cached = self.cache.read(rlat, rlon, slot, wf, wt)
                if cached is not None:
                    return cached
            raw = self.om.fetch(lat, lon, forecast_days=3)
            payload = assemble_response(
                lat=lat,
                lon=lon,
                raw=raw,
                window_from=window_from,
                window_to=window_to,
                slot=slot,
            )
            self.cache.write(rlat, rlon, slot, wf, wt, payload)

        logger.info(
            "Open-Meteo extract %.0f ms | %s heures",
            (time.perf_counter() - t0) * 1000,
            len(payload.get("hours", [])),
        )
        return payload


def assemble_response(
    *,
    lat: float,
    lon: float,
    raw: dict[str, Any],
    window_from: datetime,
    window_to: datetime,
    slot: str,
) -> dict[str, Any]:
    hourly: dict[str, Any] = raw.get("hourly") or {}
    times_raw = hourly.get("time") or []
    elev = _f(raw.get("elevation")) or 0.0
    nearest_lat = _f(raw.get("latitude"))
    nearest_lon = _f(raw.get("longitude"))
    warnings: list[str] = [
        "source=open-meteo models=arome_france (0.025°). Pas de seamless, pas de HD, pas d'ARPEGE.",
        "runInitUtc = slot cache 3 h Open-Meteo, pas l'heure d'init du run Météo-France.",
        "Td isobare : Open-Meteo le dérive de T+HU. Td AGL hors 2 m : null (non publié).",
        CLOUD_BASE_RULE,
    ]

    hours: list[HourBlock] = []
    for i, ts in enumerate(times_raw):
        when = _parse_om_time(str(ts))
        if when < window_from or when > window_to:
            continue
        heights = _height_points(hourly, i, elev)
        isobars = _isobar_points(hourly, i)
        profile = fuse_profile(heights, isobars, elev)
        surface = SurfaceBlock(
            t2m=_r1(_f(_at(hourly, "temperature_2m", i))),
            rh2m=_ri(_f(_at(hourly, "relative_humidity_2m", i))),
            wind10=_ri(_f(_at(hourly, "wind_speed_10m", i))),
            dir10=_ri(_f(_at(hourly, "wind_direction_10m", i))),
            gust10=_ri(_f(_at(hourly, "wind_gusts_10m", i))),
            cape=_ri(_f(_at(hourly, "cape", i))),
            precip=_r1(_f(_at(hourly, "precipitation", i))),
            cloudLow=_ri(_f(_at(hourly, "cloud_cover_low", i))),
            cloudMid=_ri(_f(_at(hourly, "cloud_cover_mid", i))),
            cloudHigh=_ri(_f(_at(hourly, "cloud_cover_high", i))),
            cloudBaseM=estimate_cloud_base(profile),
            psfc=_r1(_f(_at(hourly, "surface_pressure", i))),
        )
        hours.append(HourBlock(time=when, surface=surface, profile=profile))

    return {
        "model": "AROME",
        "grid": OM_GRID,
        "source": "open-meteo",
        "openMeteoModel": OM_MODEL,
        "lat": lat,
        "lon": lon,
        "modelElevationM": int(round(elev)),
        "nearestCell": {
            "lat": nearest_lat if nearest_lat is not None else round_point(lat, lon)[0],
            "lon": nearest_lon if nearest_lon is not None else round_point(lat, lon)[1],
        },
        "runInitUtc": slot,
        "runAvailable": True,
        "timezone": "Europe/Paris",
        "fetchedAt": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "hours": [h.as_dict("Europe/Paris") for h in hours],
        "warnings": warnings,
        "cloudBaseRule": CLOUD_BASE_RULE,
        "attribution": (
            "Données Météo-France AROME via Open-Meteo (CC BY 4.0). "
            "Licence Ouverte 2.0 / Météo-France. "
            "https://open-meteo.com/en/licence"
        ),
    }


def _at(hourly: dict[str, Any], key: str, index: int) -> Any:
    series = hourly.get(key)
    if not isinstance(series, list) or index >= len(series):
        return None
    return series[index]


def _parse_om_time(value: str) -> datetime:
    text = value.strip()
    if text.endswith("Z"):
        text = text[:-1] + "+00:00"
    dt = datetime.fromisoformat(text)
    if dt.tzinfo is None:
        return dt.replace(tzinfo=TIMEZONE)
    return dt


def _height_points(hourly: dict[str, Any], index: int, elev: float) -> list[ProfilePoint]:
    points: list[ProfilePoint] = []
    t2 = _f(_at(hourly, "temperature_2m", index))
    td2 = _f(_at(hourly, "dew_point_2m", index))
    rh2 = _f(_at(hourly, "relative_humidity_2m", index))
    w10 = _f(_at(hourly, "wind_speed_10m", index))
    d10 = _f(_at(hourly, "wind_direction_10m", index))
    points.append(
        ProfilePoint(
            z=int(round(elev + 2)),
            src="h2",
            t=_r1(t2),
            rh=_ri(rh2),
            td=_r1(td2),
            wind=_ri(w10),
            dir=_ri(d10),
            cloud=None,
        )
    )
    for h in HEIGHT_AGL_M:
        t = _f(_at(hourly, f"temperature_{h}m", index))
        wind = _f(_at(hourly, f"wind_speed_{h}m", index))
        direc = _f(_at(hourly, f"wind_direction_{h}m", index))
        if all(v is None for v in (t, wind, direc)):
            continue
        points.append(
            ProfilePoint(
                z=int(round(elev + h)),
                src=f"h{h}",
                t=_r1(t),
                rh=None,
                td=None,
                wind=_ri(wind),
                dir=_ri(direc),
                cloud=None,
            )
        )
    return points


def _isobar_points(hourly: dict[str, Any], index: int) -> list[ProfilePoint]:
    points: list[ProfilePoint] = []
    for p in PRESSURE_HPA:
        z = _f(_at(hourly, f"geopotential_height_{p}hPa", index))
        if z is None:
            continue
        t = _f(_at(hourly, f"temperature_{p}hPa", index))
        td = _f(_at(hourly, f"dew_point_{p}hPa", index))
        rh = _f(_at(hourly, f"relative_humidity_{p}hPa", index))
        cloud = _f(_at(hourly, f"cloud_cover_{p}hPa", index))
        wind = _f(_at(hourly, f"wind_speed_{p}hPa", index))
        direc = _f(_at(hourly, f"wind_direction_{p}hPa", index))
        points.append(
            ProfilePoint(
                z=int(round(z)),
                src=f"p{p}",
                t=_r1(t),
                rh=_ri(rh),
                td=_r1(td),
                wind=_ri(wind),
                dir=_ri(direc),
                cloud=_ri(cloud),
            )
        )
    return points


def resolve_window(
    time_from: datetime | None,
    time_to: datetime | None,
    now: datetime | None = None,
) -> tuple[datetime, datetime]:
    from datetime import timedelta

    now_paris = (now or datetime.now(TIMEZONE)).astimezone(TIMEZONE)
    default_from = now_paris.replace(hour=7, minute=0, second=0, microsecond=0)
    default_to = (now_paris + timedelta(days=2)).replace(
        hour=22, minute=0, second=0, microsecond=0
    )
    start = time_from.astimezone(TIMEZONE) if time_from else default_from
    end = time_to.astimezone(TIMEZONE) if time_to else default_to
    if end <= start:
        raise ValueError("Paramètre to doit être postérieur à from")
    return start, end

File: frontend/src/api/pipeline.ts
import { HEIGHT_AGL_M, OM_GRID, OM_MODEL, PRESSURE_HPA, type OpenMeteoRaw } from "./openmeteo";
import type { AromeResponse, Hour, ProfilePoint, Surface } from "./types";

export const MERGE_MIN_DZ_M = 80;
export const PROFILE_Z_MAX_M = 6000;
export const CLOUD_BASE_RH = 90;
export const CLOUD_BASE_FRACT = 50;

export const CLOUD_BASE_RULE =
  `cloudBaseM = AMSL of the first profile level (z ascending) where ` +
  `RH >= ${CLOUD_BASE_RH} % or CLD_FRACT >= ${CLOUD_BASE_FRACT} %; ` +
  "null otherwise. No interpolation.";

const PARIS = "Europe/Paris";

const PARTS_FMT = new Intl.DateTimeFormat("en-US", {
  timeZone: PARIS,
  year: "numeric",
  month: "2-digit",
  day: "2-digit",
  hour: "2-digit",
  minute: "2-digit",
  hour12: false,
});

const OFFSET_FMT = new Intl.DateTimeFormat("en-US", {
  timeZone: PARIS,
  timeZoneName: "longOffset",
});

type WallParts = { y: number; m: number; d: number; h: number; min: number };

function parisParts(ms: number): WallParts {
  const parts = PARTS_FMT.formatToParts(new Date(ms));
  const get = (type: Intl.DateTimeFormatPartTypes): number =>
    Number(parts.find((p) => p.type === type)?.value ?? "0");
  const h24 = get("hour");
  return {
    y: get("year"),
    m: get("month"),
    d: get("day"),
    h: h24 === 24 ? 0 : h24,
    min: get("minute"),
  };
}

function parisOffsetMs(pseudoUtcMs: number): number {
  const name =
    OFFSET_FMT.formatToParts(new Date(pseudoUtcMs))
      .find((p) => p.type === "timeZoneName")
      ?.value ?? "GMT+00:00";
  const m = /GMT([+-])(\d{1,2})(?::(\d{2}))?/.exec(name);
  if (!m) return 0;
  const sign = m[1] === "-" ? -1 : 1;
  return sign * (Number(m[2]) * 60 + Number(m[3] ?? "0")) * 60000;
}

const pad = (n: number): string => String(n).padStart(2, "0");

function naiveISO(p: WallParts, h: number, min: number): string {
  return `${p.y}-${pad(p.m)}-${pad(p.d)}T${pad(h)}:${pad(min)}`;
}

function offsetString(deltaMs: number): string {
  const total = Math.round(Math.abs(deltaMs) / 60000);
  const sign = deltaMs < 0 ? "-" : "+";
  return `${sign}${pad(Math.floor(total / 60))}:${pad(total % 60)}`;
}

function parseParisWall(value: string): { instant: number; offsetMs: number } | null {
  const m = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2})/.exec(value.trim());
  if (!m) return null;
  const pseudo = Date.UTC(
    Number(m[1]),
    Number(m[2]) - 1,
    Number(m[3]),
    Number(m[4]),
    Number(m[5]),
  );
  const offsetMs = parisOffsetMs(pseudo);
  return { instant: pseudo - offsetMs, offsetMs };
}

export function cacheSlotUtc(nowMs: number = Date.now()): string {
  const d = new Date(nowMs);
  d.setUTCMinutes(0, 0, 0);
  d.setUTCHours(d.getUTCHours() - (d.getUTCHours() % 3));
  return `${d.toISOString().slice(0, 19)}Z`;
}

export type TimeWindow = {
  start: number;
  end: number;
  fromISO: string;
  toISO: string;
};

export function resolveWindow(nowMs: number = Date.now()): TimeWindow {
  const startParts = parisParts(nowMs);
  const startPseudo = Date.UTC(startParts.y, startParts.m - 1, startParts.d, 7, 0);
  const start = startPseudo - parisOffsetMs(startPseudo);
  const endParts = parisParts(nowMs + 2 * 86_400_000);
  const endPseudo = Date.UTC(endParts.y, endParts.m - 1, endParts.d, 22, 0);
  const end = endPseudo - parisOffsetMs(endPseudo);
  return {
    start,
    end,
    fromISO: `${naiveISO(startParts, 7, 0)}:00${offsetString(startPseudo - start)}`,
    toISO: `${naiveISO(endParts, 22, 0)}:00${offsetString(endPseudo - end)}`,
  };
}

export function roundPoint(lat: number, lon: number): [number, number] {
  const r = (v: number): number => Number((Math.round(v / 0.01) * 0.01).toFixed(4));
  return [r(lat), r(lon)];
}

function num(v: unknown): number | null {
  if (v == null) return null;
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
}

const r1 = (x: number | null): number | null => (x == null ? null : Math.round(x * 10) / 10);
const ri = (x: number | null): number | null => (x == null ? null : Math.round(x));

type Hourly = Record<string, Array<number | string | null>>;

function at(hourly: Hourly, key: string, index: number): number | null {
  const series = hourly[key];
  if (!Array.isArray(series) || index >= series.length) return null;
  return num(series[index]);
}

function heightPoints(hourly: Hourly, i: number, elev: number): ProfilePoint[] {
  const points: ProfilePoint[] = [
    {
      z: Math.round(elev + 2),
      src: "h2",
      t: r1(at(hourly, "temperature_2m", i)),
      rh: ri(at(hourly, "relative_humidity_2m", i)),
      td: r1(at(hourly, "dew_point_2m", i)),
      wind: ri(at(hourly, "wind_speed_10m", i)),
      dir: ri(at(hourly, "wind_direction_10m", i)),
      cloud: null,
    },
  ];
  for (const h of HEIGHT_AGL_M) {
    const t = at(hourly, `temperature_${h}m`, i);
    const wind = at(hourly, `wind_speed_${h}m`, i);
    const dir = at(hourly, `wind_direction_${h}m`, i);
    if (t == null && wind == null && dir == null) continue;
    points.push({
      z: Math.round(elev + h),
      src: `h${h}`,
      t: r1(t),
      rh: null,
      td: null,
      wind: ri(wind),
      dir: ri(dir),
      cloud: null,
    });
  }
  return points;
}

function isobarPoints(hourly: Hourly, i: number): ProfilePoint[] {
  const points: ProfilePoint[] = [];
  for (const p of PRESSURE_HPA) {
    const z = at(hourly, `geopotential_height_${p}hPa`, i);
    if (z == null) continue;
    points.push({
      z: Math.round(z),
      src: `p${p}`,
      t: r1(at(hourly, `temperature_${p}hPa`, i)),
      rh: ri(at(hourly, `relative_humidity_${p}hPa`, i)),
      td: r1(at(hourly, `dew_point_${p}hPa`, i)),
      wind: ri(at(hourly, `wind_speed_${p}hPa`, i)),
      dir: ri(at(hourly, `wind_direction_${p}hPa`, i)),
      cloud: ri(at(hourly, `cloud_cover_${p}hPa`, i)),
    });
  }
  return points;
}

export function fuseProfile(
  heights: ProfilePoint[],
  isobars: ProfilePoint[],
  modelElev: number,
): ProfilePoint[] {
  const keptIso = isobars.filter(
    (pt) =>
      pt.z >= modelElev && !heights.some((h) => Math.abs(pt.z - h.z) < MERGE_MIN_DZ_M),
  );
  return [...heights, ...keptIso]
    .sort((a, b) => a.z - b.z)
    .filter((p) => p.z <= PROFILE_Z_MAX_M);
}

export function estimateCloudBase(profile: ProfilePoint[]): number | null {
  for (const pt of profile) {
    if ((pt.rh != null && pt.rh >= CLOUD_BASE_RH) || (pt.cloud != null && pt.cloud >= CLOUD_BASE_FRACT)) {
      return pt.z;
    }
  }
  return null;
}

function surfaceOf(hourly: Hourly, i: number, profile: ProfilePoint[]): Surface {
  return {
    t2m: r1(at(hourly, "temperature_2m", i)),
    rh2m: ri(at(hourly, "relative_humidity_2m", i)),
    wind10: ri(at(hourly, "wind_speed_10m", i)),
    dir10: ri(at(hourly, "wind_direction_10m", i)),
    gust10: ri(at(hourly, "wind_gusts_10m", i)),
    cape: ri(at(hourly, "cape", i)),
    precip: r1(at(hourly, "precipitation", i)),
    cloudLow: ri(at(hourly, "cloud_cover_low", i)),
    cloudMid: ri(at(hourly, "cloud_cover_mid", i)),
    cloudHigh: ri(at(hourly, "cloud_cover_high", i)),
    cloudBaseM: estimateCloudBase(profile),
    psfc: r1(at(hourly, "surface_pressure", i)),
  };
}

export function assembleResponse(
  lat: number,
  lon: number,
  raw: OpenMeteoRaw,
  slot: string,
  window: TimeWindow,
): AromeResponse {
  const hourly: Hourly = raw.hourly ?? {};
  const times = (hourly.time ?? []) as string[];
  const elev = num(raw.elevation) ?? 0;
  const [rlat, rlon] = roundPoint(lat, lon);
  const warnings = [
    "source=open-meteo models=arome_france (0.025°). No seamless, no HD, no ARPEGE.",
    "runInitUtc = Open-Meteo 3 h cache slot, not the Météo-France run init time.",
    "Isobaric Td: derived by Open-Meteo from T+RH. AGL Td beyond 2 m: null (not published).",
    CLOUD_BASE_RULE,
  ];

  const hours: Hour[] = [];
  for (let i = 0; i < times.length; i++) {
    const parsed = parseParisWall(times[i] ?? "");
    if (!parsed) continue;
    if (parsed.instant < window.start || parsed.instant > window.end) continue;
    const profile = fuseProfile(
      heightPoints(hourly, i, elev),
      isobarPoints(hourly, i),
      elev,
    );
    hours.push({
      time: `${times[i]}:00${offsetString(parsed.offsetMs)}`,
      surface: surfaceOf(hourly, i, profile),
      profile,
    });
  }

  return {
    model: "AROME",
    grid: OM_GRID,
    source: "open-meteo",
    openMeteoModel: OM_MODEL,
    lat,
    lon,
    modelElevationM: Math.round(elev),
    nearestCell: { lat: num(raw.latitude) ?? rlat, lon: num(raw.longitude) ?? rlon },
    runInitUtc: slot,
    runAvailable: true,
    timezone: PARIS,
    fetchedAt: new Date().toISOString().slice(0, 19) + "Z",
    hours,
    warnings,
    cloudBaseRule: CLOUD_BASE_RULE,
    attribution:
      "Météo-France AROME data via Open-Meteo (CC BY 4.0). " +
      "Licence Ouverte 2.0 / Météo-France. " +
      "https://open-meteo.com/en/licence",
  };
}
